Added drop_columns feature; usefull for optimizing cutoffs

monk-04 opened the pull request:
If an HDF5 file with larger radial cutoffs and uniform knot spacing exists, it is possible to fit a model with smaller cutoffs without regenerating the HDF5 file. This feature can be used for optimizing the radial cutoffs.

This pull request implements the above feature.

regression/optimize.py provides some useful functions for getting the appropriate columns. These columns correspond to the columns of the HDF5 file (created using larger cutoffs), which should be ignored while fitting the model to lower cutoffs.

'least_squares.WeightedLinearModel.fit_from_file' and 'least_squares.WeightedLinearModel.batched_predict' accepts new argument 'drop_columns'. Providing a list of columns to this argument results in ignoring those columns while fitting and predicting steps.

Appropriate unit tests were added to 'tests/test_optimize.py' 

- [x] Create an example for the W system demonstrating this feature

JasonGibsonUfl reviewed (changes requested):
This is an exxcelent feature of UF3 that needs to be highlighted and easily accessible. However for it to be useful it needs to be functional for more than rmin != 0. Another point of concern is that this feature will not be apparent unless a user digs into the source code. An example notebook would benefit this feature. 

`list` in lines 365 and 490 should be `List[str]`.

The function `get_bspline_config` is too rigid and variables are defined in arbitrary locations. I worry this may lead to incorrect use. For example `trailing_trim` and `leading_trim` are defined seemingly in the middle of the functions. If for some reason a user featurizes using `trailing_trim =4` they would not know that this function does not work for the given configuration.

We need to be a bit selective in how verbose we are in method and variable names. For example: `def get_possible_lower_cutoffs(original_bspline_config):` can be reduced to `def get_cuttoffs(bspline_config)` or `def get_lower_cuttoffs(bspline_config)`

In line 158 `original_bspline_config` is stated to be `bspline_config` however this is an internal definition. `original_bspline_config` should be defined as  `bspline.BSplineBasis`

method description in line 154 should be `3-body` not `2-body`

Can you provide a more informative `ValueError` for lines 101-116. It is not clear what the failed internal check is

monk-04 commented:
I have a checklist in my first comment that talks about adding an example.

Regarding the comment about the code not working when rmin!=0, the code might work when rmin!=0, I just haven't tested it. We should probably keep the condition for now and later remove it when we have extensively tested it.

I will make the changes in line 365 and 490.

I can make get_bspline_config a bit more flexible like accepting different knot spacing for 2-body and 3-body term. Regarding the leading trim and trailing trim, as I said for rmin I haven't had the time to test other settings.

I will add extra checks in `get_columns_to_drop_2b` and `get_columns_to_drop_3b`, to check for trims, knot_spacing, and some other stuff.

@JasonGibsonUfl I like your suggestion for renaming the functions.

Suggestions for line 158 and 154 noted.

The `ValueError` for internal checks is redundant check. For the drop_columns feature to work the smaller cutoff needs to be a knot of the original knots with larger cutoffs.

monk-04 commented:
As per @JasonGibsonUfl suggestions, I have made `get_bspline_config` a bit more flexible. It now works when `rmin!=0`.

I added some extra checks to `get_columns_to_drop_2b` and `get_columns_to_drop_3b` functions.

Also added one more unit test (see `config_2` in `test_optimize.py`), which is failing with my current `get_columns_to_drop_3b` algorithm.

The algorithm only fails when the `modify_3b_cutoff` is either 4.5 or 6.3. It works for other values like for example 5.4, 7.2, 1.8, etc.

@JasonGibsonUfl @sxie22 any suggestions?

monk-04 commented:
The algorithm is only failing when `rmin_3b!=0`. When `rmin_3b=0`, it passes all the cases I was able to come up with.

monk-04 commented:
I think I have found the reason why my `drop_column` algorithm is failing. Refer to issue #89 for more details.

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    if not (np.isclose((rmax_2b-rmin_2b)%knot_spacing_2b,knot_spacing_2b) or \
+            np.isclose((rmax_2b-rmin_2b)%knot_spacing_2b,0)):
+
+            raise ValueError("Provided rmax_2b does not conatin integer number of \n\
```
This line is line seems to be over-indented.

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    if not (np.isclose((rmax_3b-rmin_3b)%knot_spacing_3b,knot_spacing_3b) or \
+            np.isclose((rmax_3b-rmin_3b)%knot_spacing_3b,0)):
+
+            raise ValueError("Provided rmax_3b does not conatin integer number of \n\
```
This one as well.

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+            raise ValueError("Provided rmax_3b does not conatin integer number of \n\
+                knots, seperated by knot_spacing_3b")
+
+    if leading_trim!=0:
```
missing whitespace around operator

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    if leading_trim!=0:
+        raise ValueError("Currrent version is only tested for leading_trim=0")
+
+    if trailing_trim!=3:
```
missing whitespace around operator

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+        raise ValueError("Currrent version is only tested for trailing_trim=3")
+
+    rmax_3b_double = rmax_3b*2
+    if not(np.isclose((rmax_3b_double%knot_spacing_3b),0) or \
```
missing whitespace around modulo operator

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    reso_3b_double = round((rmax_3b_double - rmin_3b)/knot_spacing_3b)
+    
+    r_min_map = {i:rmin_2b for i in chemical_system.interactions_map[2]}
+    r_min_map.update({i:[rmin_3b,rmin_3b,rmin_3b] for i in chemical_system.interactions_map[3]})
```
missing whitespace after ','

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    rmax_3b_poss = original_bspline_config.knots_map[interaction_3b][0][4:-3]#[1::2]
+
+    for i in range(rmax_2b_poss.shape[0]):
+        if rmax_2b_poss[i] not in  original_bspline_config.knots_map[interaction_2b]:
```
Extra white space after `in`

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    #        print(interaction_3b,rmax_3b_poss[i],2*rmax_3b_poss[i])
+    #        raise ValueError("Internal check failed-->3B_2!!")
+
+    return {"rmax_2b_poss":rmax_2b_poss, "rmax_3b_poss":rmax_3b_poss}
```
missing whitespace after ':'

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+        columns_to_drop_2b (list): Should be passed to drop_columns argument of
+            fit_from_file
+    """
+    if original_bspline_config.leading_trim!=0:
```
missing whitespace around operator

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+
+    columns_to_drop_3b = []
+    for interaction in original_bspline_config.interactions_map[3]:
+        if not modify_3b_cutoff in original_bspline_config.knots_map[interaction][0]:
```
This is an unconventional use of the `not` operator. typically the syntax is `not in`

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+
+        columns_to_keep = grid_c[grid_c[:] != b'0000000000000000'].astype('<U16')
+
+        columns_to_keep_index_3b = np.where(np.isin(column_names[start_ind_3b:end_ind_3b],columns_to_keep))[0]
```
This variable is not used.

JasonGibsonUfl commented on uf3/regression/optimize.py:
```diff
@@ -0,0 +1,241 @@
+    return bspline_config
+
+
+def get_possible_lower_cutoffs(original_bspline_config):
```
I do not like this function name. I suggest removing `possible` as it makes the function's use confusing.

JasonGibsonUfl reviewed (commented):
I like the changes made to this feature. My comments are mostly about styling. Most are reoccurring such as no white space after `,`.  I would suggest using `flake8` to catch these.

The other concern I have is in variable naming. For example, line 229 exceeds PEP8 line length convention by over 30 characters. This is because of the length of the variable `num_columns_to_drop_3b_double` this could be shortened to something like `n_drop_3b_long`. Additionally, abbreviations should be used consistently. In line 237 index is abbreviated to `ind` but in the proceeding line it is not abbreviated.

monk-04 commented:
Formated the files according to @JasonGibsonUfl suggestions and also added an example jupyter notebook

monk-04 commented:
Merging!
